js_modules/dagster-ui/packages/ui-core/src/hooks/useQueryAndLocalStoragePersistedState.tsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+
+import {QueryPersistedDataType, useQueryPersistedState} from './useQueryPersistedState';
+import {useSetStateUpdateCallback} from './useSetStateUpdateCallback';
+
+/**
+ *
+ * Use URL query string as main source of truth with localStorage as the backup if no state is found in the query string
+ * Syncs changes back to localStorage but relies solely on queryString after the initial render
+ * @returns
+ */
+export const useQueryAndLocalStoragePersistedState = <T extends QueryPersistedDataType>(
+  props: Parameters<typeof useQueryPersistedState<T>>[0] & {
+    localStorageKey: string;
+    isEmptyState: (state: T) => boolean;
+  },
+): [T, (setterOrState: React.SetStateAction<T>) => void] => {
+  // Grab state from localStorage as "initialState"
+  const initialState = React.useMemo(() => {
+    try {
+      const value = localStorage.getItem(props.localStorageKey);
+      if (value) {
+        return props.decode?.(JSON.parse(value));
+      }
+    } catch {}
+    return undefined;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [props.localStorageKey]);
+
+  const [state, setter] = useQueryPersistedState(props);
+
+  React.useEffect(() => {
+    // If there was state in local storage and the state is currently empty then use the state
+    // we retrieved from localStorage
+    if (initialState && props.isEmptyState(state)) {
```

Comment:
I'm not sure we should define isEmptyState as a new prop. The `useQueryPersistedState` hook has a `defaults` value, and the query parameter is automatically removed if you set the value to something that matches defaults. (so if the default is `{tab: 'assets'}` and you set the value to something that encodes to that value, the ?tab query param is removed from the URL).

I think that's sort of the same as an empty state, maybe we could re-use the logic here and say `if (initialState && !isEqual(props.defaults, props.encode ? props.encode(state) : state)`?

Reply:
Hmm, I added the prop because I'm using a Set which can't really be serialized for comparison purposes. I imagine if we use more complex types (eg a class) we would need some way to tell if its empty or not.